src/CSModule.sol

```diff
@@ -149,104 +150,58 @@ contract CSModule is
-        bytes calldata publicKeys,
-        bytes calldata signatures,
-        NodeOperatorManagementProperties calldata managementProperties,
-        ICSAccounting.PermitInput calldata permit,
-        bytes32[] calldata eaProof,
-        address referrer
-    ) external whenResumed {
-        (uint256 nodeOperatorId, uint256 curveId) = _createNodeOperator(
-            managementProperties,
-            referrer,
-            eaProof
-        );
+        no.managerAddress = managementProperties.managerAddress == address(0)
+            ? from
+            : managementProperties.managerAddress;
+        no.rewardAddress = managementProperties.rewardAddress == address(0)
+            ? from
+            : managementProperties.rewardAddress;
+        if (managementProperties.extendedManagerPermissions)
+            no.extendedManagerPermissions = managementProperties
+                .extendedManagerPermissions;
 
-        uint256 amount = accounting.getBondAmountByKeysCount(
-            keysCount,
-            curveId
-        );
-        accounting.depositStETH(msg.sender, nodeOperatorId, amount, permit);
+        unchecked {
+            ++_nodeOperatorsCount;
+        }
 
-        _addKeysAndUpdateDepositableValidatorsCount(
+        emit NodeOperatorAdded(
             nodeOperatorId,
-            keysCount,
-            publicKeys,
-            signatures
+            no.managerAddress,
+            no.rewardAddress
         );
-    }
 
-    /// @inheritdoc ICSModule
-    function addNodeOperatorWstETH(
-        uint256 keysCount,
-        bytes calldata publicKeys,
-        bytes calldata signatures,
-        NodeOperatorManagementProperties calldata managementProperties,
-        ICSAccounting.PermitInput calldata permit,
-        bytes32[] calldata eaProof,
-        address referrer
-    ) external whenResumed {
-        (uint256 nodeOperatorId, uint256 curveId) = _createNodeOperator(
-            managementProperties,
-            referrer,
-            eaProof
-        );
-
-        uint256 amount = accounting.getBondAmountByKeysCountWstETH(
-            keysCount,
-            curveId
-        );
-        accounting.depositWstETH(msg.sender, nodeOperatorId, amount, permit);
-
-        _addKeysAndUpdateDepositableValidatorsCount(
-            nodeOperatorId,
-            keysCount,
-            publicKeys,
-            signatures
-        );
+        if (referrer != address(0)) emit ReferrerSet(nodeOperatorId, referrer);
+        _incrementModuleNonce();
     }
 
     /// @inheritdoc ICSModule
     function addValidatorKeysETH(
+        address from,
         uint256 nodeOperatorId,
         uint256 keysCount,
         bytes calldata publicKeys,
         bytes calldata signatures
     ) external payable whenResumed {
-        _onlyNodeOperatorManager(nodeOperatorId);
+        if (from == msg.sender) {
+            _onlyNodeOperatorManager(nodeOperatorId, from);
+        } else {
+            _checkRole(CREATE_NODE_OPERATOR_ROLE);
```

Comment:
For security reasons we can add a check for totalKeys == 0 to disallow uploading new keys to the exiting operators by gates